modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/RebalanceStatisticsUtils.java

```diff
@@ -79,536 +63,248 @@ private RebalanceStatisticsUtils() {
+    ) {
+        SB sb = new SB();
 
-        StringJoiner joiner = new StringJoiner(" ");
+        sb.a("Rebalance information per cache group (").a(suc ? "successful" : "interrupted").a(" rebalance): [")
+            .a(grpInfo(cacheGrpCtx)).a(", ").a(time(stat.start(), stat.end())).a(", restarted=")
+            .a(stat.attempt() - 1).a("] ");
 
-        if (finish)
-            writeTotalRebalanceStatistics(rebFutrs, nodeAliases, joiner);
+        Map<ClusterNode, SupplierRebalanceStatistics> supStats = stat.supplierStatistics();
+        if (supStats.isEmpty())
+            return sb.toString();
 
-        writeCacheGroupsRebalanceStatistics(rebFutrs, nodeAliases, finish, joiner);
-        writeAliasesRebalanceStatistics("p - partitions, e - entries, b - bytes, d - duration", nodeAliases, joiner);
-        writePartitionsDistributionRebalanceStatistics(rebFutrs, nodeAliases, nodeCnt, joiner);
+        sb.a(SUP_STAT_HEAD);
 
-        return joiner.toString();
-    }
+        int nodeId = 0;
+        for (SupplierRebalanceStatistics supStat : supStats.values()) {
+            long fp = supStat.fullParts(), hp = supStat.histParts();
+            long fe = supStat.fullEntries(), he = supStat.histEntries();
+            long fb = supStat.fullBytes(), hb = supStat.histBytes();
 
-    /**
-     * Write total statistics for rebalance.
-     *
-     * @param rebFutrs Participating in successful and not rebalances, require not null.
-     * @param nodeAliases For print nodeId=1 instead long string, require not null.
-     * @param joiner For write statistics, require not null.
-     */
-    private static void writeTotalRebalanceStatistics(
-        final Map<CacheGroupContext, Collection<RebalanceFuture>> rebFutrs,
-        final Map<ClusterNode, Integer> nodeAliases,
-        final StringJoiner joiner
-    ) {
-        assert nonNull(rebFutrs);
-        assert nonNull(nodeAliases);
-        assert nonNull(joiner);
+            sb.a(supInfo(nodeId++, fp, hp, fe, he, fb, hb, supStat.start(), supStat.end()));
+        }
 
-        long minStartTime = minStartTime(toRebalanceFutureStream(rebFutrs));
-        long maxEndTime = maxEndTime(toRebalanceFutureStream(rebFutrs));
+        sb.a(SUP_STAT_ALIASES);
 
-        joiner.add("Total information (" + SUCCESSFUL_OR_NOT_REBALANCE_TEXT + "):")
-            .add("[" + toStartEndDuration(minStartTime, maxEndTime) + "]");
+        nodeId = 0;
+        for (ClusterNode supNode : supStats.keySet())
+            sb.a(supInfo(nodeId++, supNode));
 
-        Map<ClusterNode, List<RebalanceMessageStatistics>> supplierStat =
-            toSupplierStatistics(toRebalanceFutureStream(rebFutrs));
+        if (!availablePrintPartitionsDistribution(stat.end() - stat.start()))
+            return sb.toString();
 
-        writeSupplierRebalanceStatistics(supplierStat, nodeAliases, joiner);
-    }
+        sb.a("Partitions distribution per cache group (").a(suc ? "successful" : "interrupted").a(" rebalance): [")
```

Comment:
Let's also include to log message insight on why this extended information is printed.
E. g. Rebalance duration was greater than <actual value> minutes, printing detailed information about partitions distribution (threshold can be changed by setting <property name>).